README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->


## 🪄 Swipe effect
A simple widget helps with adding an effect to your swipe actions.<br>
You can use it by wrapping your widget with <b>SwipeEffect</b> and listen to the swipes.


## ✨ Usage


```dart
SwipeEffect(
      direction: TextDirection.rtl, 
      color: Colors.green.withAlpha(70),
      verticalTolerance: 1.0,
      startDeltaPx: 30,
      callbackDeltaRatio: 0.25,
      callback: () {
        // Navigator.pop(context);
      },
      child: ...,
    );
```
## 🔮 Showcase
![image](https://github.com/shadi-danhash/swipe_effect/blob/main/images/gif1.gif?raw=true)
## ⚙️ Parameters

| Parameter | Description |
|---|---|
| `child`* | Your child widget. |
| `callback`* | A call back will be triggered after the swipe action. |
| `color`| Color of the effect curve. |
| `direction` | direction of the swipe `TextDirection.ltr` or`TextDirection.rtl` |
| `verticalTolerance`| Double value determines a ratio from the screen height that after swiping in vertical axis the effect will be dismissed. |
| `callbackDeltaRatio`| Double value determines a ratio from the screen width to trigger the callback after reaching it.|
| `startDeltaPx`|Determines where to start the effect from the beginning of the widget (in pixel) |



## 💻 Authors

- [Shadi Danhash](https://github.com/shadi-danhash)
- [Pierre Baba](https://github.com/PierreBaba)


CHANGELOG.md
## 0.0.1

* initial release.


## 1.0.1

* More customizable features
* ability to control when to trigger the listener
* ability to use left,right directions
* More design effects

## 1.0.2

* Add documentation

example/lib/main.dart
import 'package:example/screens/first_exapmle.dart';
import 'package:example/screens/seconde_example.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      routes: {
        "demo": (ctx) => const Demo(),
        "first": (ctx) => const FirstExample(),
        "second": (ctx) => const SecondExample(),
      },
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      initialRoute: "demo",
    );
  }
}

class Demo extends StatelessWidget {
  const Demo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Swipe Effect Demo')),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            MaterialButton(
              color: Colors.blueAccent,
              onPressed: () => Navigator.of(context).pushNamed("first"),
              padding: const EdgeInsets.all(50),
              child: const Text(
                "Example 1",
                style: TextStyle(color: Colors.white, fontSize: 24),
              ),
              highlightColor: Colors.transparent,
              splashColor: Colors.blueAccent.shade100,
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(10)),
            ),
            const SizedBox(
              height: 20,
            ),
            MaterialButton(
              color: Colors.greenAccent,
              highlightColor: Colors.transparent,
              onPressed: () => Navigator.of(context).pushNamed("second"),
              padding: const EdgeInsets.all(50),
              child: const Text(
                "Example 2",
                style: TextStyle(color: Colors.white, fontSize: 24),
              ),
              splashColor: Colors.greenAccent.shade100,
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(10)),
            ),
          ],
        ),
      ),
    );
  }
}


lib/swipe_effect.dart
library swipe_effect;

import 'package:flutter/material.dart';

class SwipeEffect extends StatefulWidget {
  final Color color;
  final Widget child;
  final TextDirection direction;
  final double verticalTolerance;
  final double callbackDeltaRatio;
  final double startDeltaPx;

  ///
  /// SwipeEffect constructor
  /// [child] Your child widget.
  /// [callback] A call back will be triggered after the swipe action.
  /// [color] Color of the effect curve.
  /// [direction] direction of the swipe `TextDirection.ltr` or`TextDirection.rtl`
  /// [verticalTolerance] Double value determines a ratio from the screen height that after swiping in vertical axis the effect will be dismissed.
  /// [callbackDeltaRatio] Double value determines a ratio from the screen width to trigger the callback after reaching it.
  /// [startDeltaPx] Determines where to start the effect from the beginning of the widget (in pixel)
  ///

  const SwipeEffect({
    Key? key,
    this.color = Colors.cyanAccent,
    required this.child,
    required this.callback,
    this.direction = TextDirection.rtl,
    this.verticalTolerance = 0.1,
    this.callbackDeltaRatio = 0.25,
    this.startDeltaPx = 30,
  }) : super(key: key);

  @override
  State<SwipeEffect> createState() => _SwipeEffectState();

  final VoidCallback callback;
}

class _SwipeEffectState extends State<SwipeEffect> {
  late Offset point;
  late Offset start;
  late Size size;
  late double startX;
  bool isDragging = false;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    size = MediaQuery.of(context).size;
    startX = widget.direction == TextDirection.rtl ? size.width : 0;
    point = Offset(startX, 0);
    start = Offset.zero;
  }

  ///
  /// Detects the swipe gesture and builds a "CustomPaint" widget based on that gesture.
  ///
  @override
  Widget build(BuildContext context) {
    return SizedBox(
        height: size.height,
        width: size.width,
        child: GestureDetector(
          onHorizontalDragStart: (details) {
            isDragging =
                (startX - details.localPosition.dx).abs() < widget.startDeltaPx;
            start = details.localPosition;
          },
          onHorizontalDragEnd: (details) {
            var tt = (startX - point.dx).abs();
            if (isDragging && tt > size.width * widget.callbackDeltaRatio) {
              widget.callback();
            }
            setState(() {
              isDragging = false;
              point = Offset(startX, 0);
              start = Offset.zero;
            });
          },
          onHorizontalDragUpdate: (details) {
            if ((startX - start.dx).abs() > widget.startDeltaPx) return;
            if ((start.dy - details.localPosition.dy).abs() >
                size.height * widget.verticalTolerance) {
              setState(() {
                point = Offset(startX, 0);
                start = const Offset(0, 0);
              });
              return;
            }
            setState(() {
              point = details.globalPosition;
            });
          },
          child: CustomPaint(
            size: Size.infinite,
            foregroundPainter:
                BackEffect(point, widget.color, widget.direction),
            child: widget.child,
          ),
        ));
  }

  @override
  void didUpdateWidget(covariant SwipeEffect oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.direction != oldWidget.direction) {
      setState(() {
        startX = widget.direction == TextDirection.rtl ? size.width : 0;
        point = Offset(startX, 0);
        start = Offset.zero;
      });
    }
  }
}

class BackEffect extends CustomPainter {
  Offset dragPoint;
  Color color;
  TextDirection direction;

  BackEffect(this.dragPoint, this.color, this.direction);

  /// Draws path with the given [size].
  ///
  /// The path is filled with [color] and starts in [direction]
  /// The path is a combination of two quadratic curves, to add the swipe effect.
  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color
      ..style = PaintingStyle.fill;
    final dy = size.height / 5;
    final double startX = direction == TextDirection.rtl ? size.width : 0;
    final p0 = Offset(startX, 0);
    final p3 = dragPoint;
    final p6 = Offset(startX, size.height);
    final p1 = Offset(p0.dx, p0.dy + dy);
    final p2 = Offset(p3.dx, p3.dy - dy);
    final p4 = Offset(p3.dx, p3.dy + dy);
    final p5 = Offset(p6.dx, p6.dy - dy);
    final path = Path()
      ..moveTo(p0.dx, p0.dy)
      ..cubicTo(p1.dx, p1.dy, p2.dx, p2.dy, p3.dx, p3.dy)
      ..cubicTo(p4.dx, p4.dy, p5.dx, p5.dy, p6.dx, p6.dy)
      ..close();
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return this != oldDelegate;
  }
}


example/lib/screens/seconde_example.dart
import 'package:flutter/material.dart';
import 'package:swipe_effect/swipe_effect.dart';

class SecondExample extends StatefulWidget {
  const SecondExample({Key? key, this.title = "Second Example"})
      : super(key: key);

  final String title;

  @override
  State<SecondExample> createState() => _SecondExampleState();
}

class _SecondExampleState extends State<SecondExample> {
  TextDirection direction = TextDirection.rtl;
  double start = 30;
  double callbackDelta = 0.25;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return SwipeEffect(
      direction: direction,
      color: Colors.cyanAccent.withAlpha(70),
      verticalTolerance: 1.0,
      startDeltaPx: start,
      callbackDeltaRatio: callbackDelta,
      callback: () {
        Navigator.pop(context);
      },
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            Navigator.of(context).pushNamed("second");
          },
          child: const Icon(Icons.add),
          tooltip: "push new page",
        ),
        body: Center(
          child: Text(
            "Swipe to the right to go back\n\nOR\n\nClick on the + icon to push new page",
            style: TextStyle(
                color: Colors.tealAccent.shade400,
                fontSize: 24,
                fontWeight: FontWeight.bold),
            textAlign: TextAlign.center,
          ),
        ),
      ),
    );
  }
}


example/lib/screens/first_exapmle.dart
import 'package:flutter/material.dart';
import 'package:swipe_effect/swipe_effect.dart';

class FirstExample extends StatefulWidget {
  const FirstExample({Key? key, this.title = "First Example"})
      : super(key: key);

  final String title;

  @override
  State<FirstExample> createState() => _FirstExampleState();
}

class _FirstExampleState extends State<FirstExample> {
  TextDirection direction = TextDirection.rtl;
  double start = 30;
  double callbackDelta = 0.25;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return SwipeEffect(
      direction: direction,
      color: direction == TextDirection.rtl
          ? Colors.red.withAlpha(70)
          : Colors.green.withAlpha(70),
      verticalTolerance: 1.0,
      startDeltaPx: start,
      callbackDeltaRatio: callbackDelta,
      callback: () {
        setState(() {
          start += 10;
          callbackDelta += 0.1;
        });
        // Navigator.pop(context);
      },
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            setState(() {
              direction = direction == TextDirection.rtl
                  ? TextDirection.ltr
                  : TextDirection.rtl;
              callbackDelta = 0.25;
              start = 30;
            });
          },
          child: const Icon(Icons.swap_horiz),
        ),
        body: Stack(
          children: [
            Align(
              alignment: direction == TextDirection.ltr
                  ? Alignment.centerLeft
                  : Alignment.centerRight,
              child: Container(
                color: Colors.blue.withOpacity(0.5),
                width: start,
              ),
            ),
            Align(
              alignment: direction == TextDirection.ltr
                  ? Alignment.centerLeft
                  : Alignment.centerRight,
              child: Container(
                color: Colors.black.withAlpha(30),
                width: callbackDelta * size.width,
              ),
            ),
            Center(
              child: Text(
                direction == TextDirection.rtl ? 'Swipe left' : "Swipe right",
                style: TextStyle(
                  color: direction == TextDirection.rtl
                      ? Colors.red
                      : Colors.green,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
